python/ray/dag/compiled_dag_node.py

```diff
@@ -584,24 +615,38 @@ def prepare(self, overlap_gpu_communication: bool = False):
             overlap_gpu_communication: Whether to overlap GPU communication with
                 computation during DAG execution to improve performance
         """
+        from ray.dag.collective_node import _CollectiveOperation
+        from ray.dag.p2p_node import _P2PSendOperation, _P2PRecvOperation
+
         for typ_hint in self.input_type_hints:
             typ_hint.register_custom_serializer()
         self.output_type_hint.register_custom_serializer()
-        self.input_reader.start()
-        self.output_writer.start()
-
-        self._send_stream: Union["cp.cuda.Stream", nullcontext] = nullcontext()
-        self._recv_stream: Union["cp.cuda.Stream", nullcontext] = nullcontext()
+        if self.input_reader is not None:
+            self.input_reader.start()
+        if self.output_writer is not None:
+            self.output_writer.start()
+
+        # Convert the abstract P2P operation from scheduling to the executable P2P
+        # send/recv operation.
+        if self.requires_nccl_read:
+            assert self.nccl_ch is not None
+            self.nccl_op = _P2PRecvOperation(self.nccl_ch)
+        elif self.requires_nccl_write:
+            assert self.nccl_ch is not None
+            self.nccl_ch.ensure_registered_as_writer()
+            self.nccl_op = _P2PSendOperation(self.nccl_ch)
```

Comment:
Why do we only need to do this conversion from abstract to executable operation for P2P operations and not for collective operations?